lambda/src/index.js

```diff
@@ -1,31 +1,115 @@
+const fetch = require('node-fetch')
+const fs = require('fs')
 const newman = require('newman')
 const AWS = require('aws-sdk')
 const codedeploy = new AWS.CodeDeploy({ apiVersion: '2014-10-06', region: 'us-west-2' })
+const s3 = new AWS.S3({ apiVersion: '2014-10-06', region: 'us-west-2' })
+
+const tmpDir = process.env.TMP_DIR ? process.env.TMP_DIR : '/tmp'
 
 exports.handler = async function (event, context) {
   console.log(event)
-
   // Workaround for CodeDeploy bug.
   // Give the ALB 10 seconds to make sure the test TG has switched to the new code.
-  await sleep(10000)
+  const timer = sleep(10000)
+
+  if (process.env.POSTMAN_API_KEY) {
+    // start downloading postman files from Postman API
+    const downloadCollection = downloadFileFromPostman('collection', process.env.POSTMAN_COLLECTION_NAME)
+    const downloadEnv = downloadFileFromPostman('environment', process.env.POSTMAN_ENVIRONMENT_NAME)
+    await downloadCollection
+    await downloadEnv
+  } else {
+    // start downloading postman files from S3 Bucket
+    const downloadCollection = downloadFileFromBucket('collection', process.env.POSTMAN_COLLECTION)
+    const downloadEnv = downloadFileFromBucket('environment', process.env.POSTMAN_ENVIRONMENT)
+    await downloadCollection
+    await downloadEnv
+  }
 
   let errorFromTests
-  await runTests(process.env.POSTMAN_COLLECTION, process.env.POSTMAN_ENVIRONMENT).catch(err => { errorFromTests = err })
+  const postmanTests = runTests(
+    `${tmpDir}/collection.json`,
+    `${tmpDir}/environment.json`
+  ).catch(err => { errorFromTests = err })
+
+  await postmanTests
+  await timer
+
+  const deploymentId = event.DeploymentId
+  if (deploymentId) {
+    console.log('starting to update CodeDeploy lifecycle event hook status...')
+    const params = {
+      deploymentId: event.DeploymentId,
+      lifecycleEventHookExecutionId: event.LifecycleEventHookExecutionId,
+      status: errorFromTests ? 'Failed' : 'Succeeded'
+    }
+    try {
+      const data = await codedeploy.putLifecycleEventHookExecutionStatus(params).promise()
+      console.log(data)
+    } catch (err) {
+      console.log(err, err.stack)
+      throw err
+    }
+  } else {
+    console.log('No deployment ID found in the event. Skipping update to CodeDeploy lifecycle hook...')
+  }
+
+  if (errorFromTests) throw errorFromTests // Cause the lambda to "fail"
+}
 
-  const params = {
-    deploymentId: event.DeploymentId,
-    lifecycleEventHookExecutionId: event.LifecycleEventHookExecutionId,
-    status: errorFromTests ? 'Failed' : 'Succeeded'
+async function downloadFileFromPostman (type, name) {
+  console.log(`started download for ${tmpDir}/${type}.json`)
+  try {
+    const response = await fetch(`https://api.getpostman.com/${type}s`, {
+      method: 'GET',
+      headers: {
+        'X-Api-Key': process.env.POSTMAN_API_KEY
+      }
+    })
```

Comment:
At first, I didn't understand why you would bother getting the list of collections or environments. But I realized that the `uid`s are pretty hard to find without exporting the collections or environments.

I don't particularly love how we need to filter the list by `name` (since I have several environments just named `localhost`), but the Postman API doesn't give us any other option besides `uid`, so LGTM.

Reply:
Yeah, I agree. I guess we can make it where you can provide either the name or uid? I also thought about not requiring the environment, but haven't gotten that far yet.

Reply:
Maybe? If we find ourselves needing those things, we can make the enhancements later.

As for the `name` thing, we don't exactly have a lot of Postman collections and environments right now. The simple solution is to make environment names unique.

Reply:
>make environment names unique

I mean that I need to stop naming all of my environments `localhost`. 🙂